dace/codegen/tools/type_inference.py

```diff
@@ -0,0 +1,580 @@
+"""
+    Type inference: traverses code and returns types for all undefined symbols according to C semantics
+    infer() has a lenient implementation: if something it not inferred (for example an unsupported construct) it will not
+        return anything and it will not produce errors
+
+    This module is inspired by astunparse: https://github.com/simonpercivall/astunparse
+"""
+
+import numpy as np
+import ast
+from dace import dtypes
+from dace.codegen import cppunparse
+from dace.symbolic import SymExpr
+from dace.symbolic import symstr
+from sympy import Basic
+import sys
+
+
+def infer(code, symbols):
```

Comment:
Suggestion: split into two inferfaces:
```python
infer_types(code, symbols)  # Returns a dictionary
infer_expr_type(expr_code, symbols)  # Returns a single type
```
Quite often you are just interested in the final type of an expression, so it would be convenient to have the bottom one to save boilerplate

Reply:
Expression or assignment? E.g., `target = a+b+c ` and you want the type of target

Reply:
Both!

Reply:
yep, and I would make symbols an optional argument (`symbols=None` and then inside do `symbols = symbols or {}` to avoid weird bugs with Python default arguments)